core/src/test/java/org/ehcache/EhcacheBasicPutAllEventsTest.java

```diff
@@ -0,0 +1,216 @@
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.ehcache.event.CacheEvent;
+import org.ehcache.event.CacheEventListener;
+import org.ehcache.events.CacheEventNotificationService;
+import org.ehcache.events.CacheEventNotificationServiceImpl;
+import org.ehcache.exceptions.CacheAccessException;
+import org.ehcache.function.Function;
+import org.ehcache.spi.loaderwriter.CacheLoaderWriter;
+import org.ehcache.util.IsCreatedOrUpdated;
+import org.junit.Test;
+import org.mockito.Matchers;
+import org.mockito.Mock;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides events based testing for PUT_ALL operations
+ */
+public class EhcacheBasicPutAllEventsTest extends EhcacheBasicCrudBase {
+
+  @Mock
+  protected CacheEventListener<String, String> cacheEventListener;
+
+  protected CacheEventNotificationService<String, String> cacheEventNotificationService;
+
+  protected IsCreatedOrUpdated<String, String> isCreatedOrUpdated = new IsCreatedOrUpdated<String, String>();
+
+  @Test
+  public void testPutAllNull() throws Exception {
+    final Map<String, String> originalStoreContent = getEntryMap(KEY_SET_A, KEY_SET_B);
+    final FakeStore fakeStore = new FakeStore(originalStoreContent);
+    this.store = spy(fakeStore);
+    final Ehcache<String, String> ehcache = this.getEhcache();
+    try {
+      ehcache.putAll(null);
+      fail();
+    } catch (NullPointerException e) {
+      // Expected
+    }
+    verify(cacheEventListener, never()).onEvent(Matchers.<CacheEvent<String, String>>any());
+    try {
+      ehcache.getRuntimeConfiguration().deregisterCacheEventListener(cacheEventListener);
+      fail();
+    } catch (UnsupportedOperationException e) {
+      //expected
+    }
+  }
+
+  @Test
+  public void testPutAllNullKey() throws Exception {
+    final Map<String, String> originalStoreContent = getEntryMap(KEY_SET_A, KEY_SET_B);
+    final FakeStore fakeStore = new FakeStore(originalStoreContent);
+    this.store = spy(fakeStore);
+
+    final Map<String, String> entries = new LinkedHashMap<String, String>();
+    for (final Map.Entry<String, String> entry : getEntryMap(KEY_SET_A).entrySet()) {
+      final String key = entry.getKey();
+      entries.put(key, entry.getValue());
+      if ("keyA2".equals(key)) {
+        entries.put(null, "nullKey");
+      }
+    }
+    final Ehcache<String, String> ehcache = this.getEhcache();
+    try {
+      ehcache.putAll(entries);
+      fail();
+    } catch (NullPointerException e) {
+      // Expected
+    }
+    verify(cacheEventListener, never()).onEvent(Matchers.<CacheEvent<String, String>>any());
+    try {
+      ehcache.getRuntimeConfiguration().deregisterCacheEventListener(cacheEventListener);
```

Comment:
Indeed. Also, I really do not like the fact that this ends up throwing an exception. Will look into this.